{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a413234c",
   "metadata": {},
   "source": [
    "# Exploring EBI gene2pheno data for parser"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7806d5e",
   "metadata": {},
   "source": [
    "## Downloading data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99a94602",
   "metadata": {},
   "source": [
    "Latest data can be downloaded from https://www.ebi.ac.uk/gene2phenotype/download. But I can't figure out the actual links for the downloads. \n",
    "\n",
    "Static release files (currently 2025-02-28) can be downloaded from the FTP site http://ftp.ebi.ac.uk/pub/databases/gene2phenotype/ (in the G2P_data_downloads subfolder). It is supposed to update on a monthly basis, according to the [README](http://ftp.ebi.ac.uk/pub/databases/gene2phenotype/README). However, I notice that the FTP site is sometimes slow on my computer (Firefox browser). \n",
    "\n",
    "I'm using all disorder panels/files:\n",
    "* Cancer\n",
    "* Cardiac\n",
    "* Developmental disorders (DD)\n",
    "* Eye\n",
    "* Hearing\n",
    "* Skeletal\n",
    "* Skin"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc2ed7fe",
   "metadata": {},
   "source": [
    "## Load data into separate pandas dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "be796066",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-03-14T08:59:26.007353Z",
     "start_time": "2025-03-14T08:59:26.003729Z"
    }
   },
   "outputs": [],
   "source": [
    "## import packages for exploring here\n",
    "\n",
    "## CX: allows multiple lines of code to print from one code block\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "InteractiveShell.ast_node_interactivity = \"all\"\n",
    "\n",
    "import pathlib\n",
    "import pandas as pd\n",
    "import glob\n",
    "\n",
    "pd.options.display.max_columns = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "f75451df",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-03-17T19:56:13.973922Z",
     "start_time": "2025-03-17T19:56:13.859559Z"
    }
   },
   "outputs": [],
   "source": [
    "## Construct paths for files\n",
    "\n",
    "## adjust based on where files are stored\n",
    "base_file_path = pathlib.Path.home().joinpath('Desktop', 'EBIgene2pheno_files', 'From_Website')\n",
    "\n",
    "cancer = pd.read_csv(base_file_path.joinpath('G2P_Cancer_2025-03-13.csv'), \n",
    "                      dtype={\"gene mim\": str, \"hgnc id\": str, \"disease mim\": str})\n",
    "cardiac = pd.read_csv(base_file_path.joinpath('G2P_Cardiac_2025-03-13.csv'), \n",
    "                      dtype={\"gene mim\": str, \"hgnc id\": str, \"disease mim\": str})\n",
    "developmental = pd.read_csv(base_file_path.joinpath('G2P_DD_2025-03-13.csv'), \n",
    "                      dtype={\"gene mim\": str, \"hgnc id\": str, \"disease mim\": str})\n",
    "eye = pd.read_csv(base_file_path.joinpath('G2P_Eye_2025-03-13.csv'), \n",
    "                      dtype={\"gene mim\": str, \"hgnc id\": str, \"disease mim\": str})\n",
    "hearing = pd.read_csv(base_file_path.joinpath('G2P_Hearing loss_2025-03-13.csv'), \n",
    "                      dtype={\"gene mim\": str, \"hgnc id\": str, \"disease mim\": str})\n",
    "skeletal = pd.read_csv(base_file_path.joinpath('G2P_Skeletal_2025-03-13.csv'), \n",
    "                      dtype={\"gene mim\": str, \"hgnc id\": str, \"disease mim\": str})\n",
    "skin = pd.read_csv(base_file_path.joinpath('G2P_Skin_2025-03-13.csv'), \n",
    "                      dtype={\"gene mim\": str, \"hgnc id\": str, \"disease mim\": str})\n",
    "\n",
    "## saving for possible future use\n",
    "# unique_filename_substring = [\n",
    "#     'Cancer',\n",
    "#     'Cardiac',\n",
    "#     'DD',\n",
    "#     'Eye',\n",
    "#     'Hearing loss',\n",
    "#     'Skeletal',\n",
    "#     'Skin'\n",
    "# ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "043dccbf",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-03-14T08:38:02.470651Z",
     "start_time": "2025-03-14T08:38:02.465891Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/colleenxu/Desktop/EBIgene2pheno_files/From_Website/G2P_Skin_2025-03-13.csv\n",
      "/Users/colleenxu/Desktop/EBIgene2pheno_files/From_Website/G2P_Hearing loss_2025-03-13.csv\n",
      "/Users/colleenxu/Desktop/EBIgene2pheno_files/From_Website/G2P_Cardiac_2025-03-13.csv\n",
      "/Users/colleenxu/Desktop/EBIgene2pheno_files/From_Website/G2P_Skeletal_2025-03-13.csv\n",
      "/Users/colleenxu/Desktop/EBIgene2pheno_files/From_Website/G2P_Cancer_2025-03-13.csv\n",
      "/Users/colleenxu/Desktop/EBIgene2pheno_files/From_Website/G2P_DD_2025-03-13.csv\n",
      "/Users/colleenxu/Desktop/EBIgene2pheno_files/From_Website/G2P_Eye_2025-03-13.csv\n"
     ]
    }
   ],
   "source": [
    "## way to get all paths for all files without knowing their actual names\n",
    "for i in base_file_path.glob(\"*.csv\"):\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "833fb8d7",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-03-14T08:38:09.954402Z",
     "start_time": "2025-03-14T08:38:09.947089Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True])"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## checking that each file's columns are the same: they are\n",
    "cancer.columns == skin.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "e99aee21",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-03-17T19:35:36.398300Z",
     "start_time": "2025-03-17T19:35:36.393297Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['g2p id', 'gene symbol', 'gene mim', 'hgnc id', 'previous gene symbols',\n",
       "       'disease name', 'disease mim', 'disease MONDO', 'allelic requirement',\n",
       "       'cross cutting modifier', 'confidence', 'variant consequence',\n",
       "       'variant types', 'molecular mechanism',\n",
       "       'molecular mechanism categorisation', 'molecular mechanism evidence',\n",
       "       'phenotypes', 'publications', 'panel', 'comments',\n",
       "       'date of last review'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cardiac.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "d28971e4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-03-17T19:35:41.350637Z",
     "start_time": "2025-03-17T19:35:41.323599Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g2p id</th>\n",
       "      <th>gene symbol</th>\n",
       "      <th>gene mim</th>\n",
       "      <th>hgnc id</th>\n",
       "      <th>previous gene symbols</th>\n",
       "      <th>disease name</th>\n",
       "      <th>disease mim</th>\n",
       "      <th>disease MONDO</th>\n",
       "      <th>allelic requirement</th>\n",
       "      <th>cross cutting modifier</th>\n",
       "      <th>confidence</th>\n",
       "      <th>variant consequence</th>\n",
       "      <th>variant types</th>\n",
       "      <th>molecular mechanism</th>\n",
       "      <th>molecular mechanism categorisation</th>\n",
       "      <th>molecular mechanism evidence</th>\n",
       "      <th>phenotypes</th>\n",
       "      <th>publications</th>\n",
       "      <th>panel</th>\n",
       "      <th>comments</th>\n",
       "      <th>date of last review</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G2P03247</td>\n",
       "      <td>DSC2</td>\n",
       "      <td>125645</td>\n",
       "      <td>3036</td>\n",
       "      <td>CDHF2; DSC3</td>\n",
       "      <td>DSC2-related arrhythmogenic right ventricular ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0012506</td>\n",
       "      <td>monoallelic_autosomal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>definitive</td>\n",
       "      <td>decreased gene product level; altered gene pro...</td>\n",
       "      <td>splice_region_variant; inframe_insertion; spli...</td>\n",
       "      <td>undetermined</td>\n",
       "      <td>inferred</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>31028357; 23911551; 21636032; 33831308; 263105...</td>\n",
       "      <td>Cardiac</td>\n",
       "      <td>Expert review done on 05/01/2022; DSC2-related...</td>\n",
       "      <td>2024-03-20 09:36:09+00:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G2P03248</td>\n",
       "      <td>DSC2</td>\n",
       "      <td>125645</td>\n",
       "      <td>3036</td>\n",
       "      <td>CDHF2; DSC3</td>\n",
       "      <td>DSC2-related arrhythmogenic right ventricular ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0012506</td>\n",
       "      <td>biallelic_autosomal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>definitive</td>\n",
       "      <td>decreased gene product level; altered gene pro...</td>\n",
       "      <td>splice_region_variant; inframe_insertion; spli...</td>\n",
       "      <td>undetermined</td>\n",
       "      <td>inferred</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>31028357; 23911551; 21636032; 33831308; 263105...</td>\n",
       "      <td>Cardiac</td>\n",
       "      <td>Expert review done on 05/01/2022; DSC2-related...</td>\n",
       "      <td>2024-03-20 09:35:19+00:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G2P03249</td>\n",
       "      <td>DSG2</td>\n",
       "      <td>125671</td>\n",
       "      <td>3049</td>\n",
       "      <td>CDHF5</td>\n",
       "      <td>DSG2-related arrhythmogenic right ventricular ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0012434</td>\n",
       "      <td>monoallelic_autosomal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>definitive</td>\n",
       "      <td>decreased gene product level; altered gene pro...</td>\n",
       "      <td>inframe_insertion; splice_donor_variant; infra...</td>\n",
       "      <td>undetermined</td>\n",
       "      <td>inferred</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21636032; 33831308; 33917638; 34400560; 240707...</td>\n",
       "      <td>Cardiac</td>\n",
       "      <td>Expert review done on 05/01/2022; DSG2-related...</td>\n",
       "      <td>2024-03-20 09:40:18+00:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>G2P03250</td>\n",
       "      <td>DSG2</td>\n",
       "      <td>125671</td>\n",
       "      <td>3049</td>\n",
       "      <td>CDHF5</td>\n",
       "      <td>DSG2-related arrhythmogenic right ventricular ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0012434</td>\n",
       "      <td>biallelic_autosomal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>definitive</td>\n",
       "      <td>decreased gene product level; altered gene pro...</td>\n",
       "      <td>inframe_insertion; splice_donor_variant; infra...</td>\n",
       "      <td>undetermined</td>\n",
       "      <td>inferred</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21636032; 33831308; 33917638; 34400560; 240707...</td>\n",
       "      <td>Cardiac</td>\n",
       "      <td>Expert review done on 05/01/2022; DSG2-related...</td>\n",
       "      <td>2024-03-20 09:37:58+00:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>G2P03251</td>\n",
       "      <td>DSP</td>\n",
       "      <td>125647</td>\n",
       "      <td>3052</td>\n",
       "      <td>DPI; DPII; KPPS2; PPKS2</td>\n",
       "      <td>DSP-related arrhythmogenic right ventricular c...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0011831</td>\n",
       "      <td>monoallelic_autosomal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>definitive</td>\n",
       "      <td>decreased gene product level; altered gene pro...</td>\n",
       "      <td>inframe_insertion; splice_donor_variant; infra...</td>\n",
       "      <td>undetermined</td>\n",
       "      <td>inferred</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32372669; 23137101; 21636032; 33831308; 313199...</td>\n",
       "      <td>Cardiac</td>\n",
       "      <td>Expert review done on 05/01/2022; DSP-related ...</td>\n",
       "      <td>2024-03-20 09:42:53+00:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>G2P03252</td>\n",
       "      <td>DSP</td>\n",
       "      <td>125647</td>\n",
       "      <td>3052</td>\n",
       "      <td>DPI; DPII; KPPS2; PPKS2</td>\n",
       "      <td>DSP-related arrhythmogenic right ventricular c...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0011831</td>\n",
       "      <td>biallelic_autosomal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>definitive</td>\n",
       "      <td>decreased gene product level; altered gene pro...</td>\n",
       "      <td>inframe_insertion; splice_donor_variant; infra...</td>\n",
       "      <td>undetermined</td>\n",
       "      <td>inferred</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32372669; 23137101; 21636032; 33831308; 313199...</td>\n",
       "      <td>Cardiac</td>\n",
       "      <td>Expert review done on 05/01/2022; DSP-related ...</td>\n",
       "      <td>2024-03-20 09:42:01+00:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>G2P03262</td>\n",
       "      <td>CASQ2</td>\n",
       "      <td>114251</td>\n",
       "      <td>1513</td>\n",
       "      <td>PDIB2</td>\n",
       "      <td>CASQ2-related catecholaminergic polymorphic ve...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0012762</td>\n",
       "      <td>monoallelic_autosomal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>moderate</td>\n",
       "      <td>decreased gene product level; altered gene pro...</td>\n",
       "      <td>missense_variant; splice_region_variant; splic...</td>\n",
       "      <td>undetermined</td>\n",
       "      <td>inferred</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12386154; 16908766; 21618644; 15176429; 326936...</td>\n",
       "      <td>Cardiac</td>\n",
       "      <td>Expert review done on 19/05/2021; CASQ2 is an ...</td>\n",
       "      <td>2024-03-13 11:49:22+00:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>G2P03263</td>\n",
       "      <td>CASQ2</td>\n",
       "      <td>114251</td>\n",
       "      <td>1513</td>\n",
       "      <td>PDIB2</td>\n",
       "      <td>CASQ2-related catecholaminergic polymorphic ve...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0012762</td>\n",
       "      <td>biallelic_autosomal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>definitive</td>\n",
       "      <td>absent gene product; altered gene product stru...</td>\n",
       "      <td>missense_variant; splice_region_variant; splic...</td>\n",
       "      <td>undetermined</td>\n",
       "      <td>inferred</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12386154; 16908766; 21618644; 15176429; 326936...</td>\n",
       "      <td>Cardiac</td>\n",
       "      <td>Expert review done on 19/05/2021; CASQ2 is an ...</td>\n",
       "      <td>2022-04-25 12:45:21+00:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>G2P03285</td>\n",
       "      <td>KCNQ1</td>\n",
       "      <td>607542</td>\n",
       "      <td>6294</td>\n",
       "      <td>JLNS1; KCNA8; KCNA9; KV7.1; KVLQT1; LQT; LQT1</td>\n",
       "      <td>KCNQ1-related long QT syndrome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0100316</td>\n",
       "      <td>monoallelic_autosomal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>definitive</td>\n",
       "      <td>decreased gene product level; altered gene pro...</td>\n",
       "      <td>frameshift_variant_NMD_triggering; inframe_ins...</td>\n",
       "      <td>undetermined</td>\n",
       "      <td>inferred</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>31983240; 9020846; 23392653; 23591039; 2704115...</td>\n",
       "      <td>Cardiac</td>\n",
       "      <td>Expert review done on 16/06/2021; Pathogenic v...</td>\n",
       "      <td>2024-03-13 11:56:22+00:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>G2P03286</td>\n",
       "      <td>KCNQ1</td>\n",
       "      <td>607542</td>\n",
       "      <td>6294</td>\n",
       "      <td>JLNS1; KCNA8; KCNA9; KV7.1; KVLQT1; LQT; LQT1</td>\n",
       "      <td>KCNQ1-related long QT syndrome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0100316</td>\n",
       "      <td>biallelic_autosomal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>definitive</td>\n",
       "      <td>decreased gene product level; altered gene pro...</td>\n",
       "      <td>frameshift_variant_NMD_triggering; inframe_ins...</td>\n",
       "      <td>undetermined</td>\n",
       "      <td>inferred</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>31983240; 9020846; 23392653; 23591039; 2704115...</td>\n",
       "      <td>Cardiac</td>\n",
       "      <td>Expert review done on 16/06/2021; Pathogenic v...</td>\n",
       "      <td>2024-03-13 11:57:14+00:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>G2P03289</td>\n",
       "      <td>ALPK3</td>\n",
       "      <td>617608</td>\n",
       "      <td>17574</td>\n",
       "      <td>KIAA1330; MAK; MIDORI</td>\n",
       "      <td>ALPK3-related hypertrophic cardiomyopathy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0054838</td>\n",
       "      <td>monoallelic_autosomal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>strong</td>\n",
       "      <td>decreased gene product level; altered gene pro...</td>\n",
       "      <td>frameshift_variant_NMD_triggering; splice_dono...</td>\n",
       "      <td>undetermined</td>\n",
       "      <td>inferred</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27106955; 31074094; 32480058; 26846950; 214411...</td>\n",
       "      <td>Cardiac</td>\n",
       "      <td>Expert review done on 24/11/2021; ALPK3 pathog...</td>\n",
       "      <td>2022-04-25 12:45:22+00:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>G2P03290</td>\n",
       "      <td>ALPK3</td>\n",
       "      <td>617608</td>\n",
       "      <td>17574</td>\n",
       "      <td>KIAA1330; MAK; MIDORI</td>\n",
       "      <td>ALPK3-related hypertrophic cardiomyopathy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0054838</td>\n",
       "      <td>biallelic_autosomal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>strong</td>\n",
       "      <td>decreased gene product level; altered gene pro...</td>\n",
       "      <td>frameshift_variant_NMD_triggering; splice_dono...</td>\n",
       "      <td>undetermined</td>\n",
       "      <td>inferred</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27106955; 31074094; 32480058; 26846950; 214411...</td>\n",
       "      <td>Cardiac</td>\n",
       "      <td>Expert review done on 24/11/2021; ALPK3 pathog...</td>\n",
       "      <td>2022-04-25 12:45:22+00:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>G2P03291</td>\n",
       "      <td>DES</td>\n",
       "      <td>125660</td>\n",
       "      <td>2770</td>\n",
       "      <td>CMD1I; CSM1; CSM2; LGMD2R</td>\n",
       "      <td>DES-related myofibrillar myopathy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0011076</td>\n",
       "      <td>monoallelic_autosomal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>definitive</td>\n",
       "      <td>absent gene product; altered gene product stru...</td>\n",
       "      <td>frameshift_variant_NMD_triggering; splice_dono...</td>\n",
       "      <td>undetermined</td>\n",
       "      <td>inferred</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30681346; 19181099; 20718792; 29926427; 194333...</td>\n",
       "      <td>Cardiac</td>\n",
       "      <td>Expert review done on 12/01/2022; DES-related ...</td>\n",
       "      <td>2022-04-25 12:45:22+00:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>G2P03292</td>\n",
       "      <td>DES</td>\n",
       "      <td>125660</td>\n",
       "      <td>2770</td>\n",
       "      <td>CMD1I; CSM1; CSM2; LGMD2R</td>\n",
       "      <td>DES-related myofibrillar myopathy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0011076</td>\n",
       "      <td>biallelic_autosomal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>definitive</td>\n",
       "      <td>absent gene product; altered gene product stru...</td>\n",
       "      <td>frameshift_variant_NMD_triggering; splice_dono...</td>\n",
       "      <td>undetermined</td>\n",
       "      <td>inferred</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30681346; 19181099; 20718792; 29926427; 194333...</td>\n",
       "      <td>Cardiac</td>\n",
       "      <td>Expert review done on 12/01/2022; DES-related ...</td>\n",
       "      <td>2024-03-26 10:31:28+00:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>G2P03446</td>\n",
       "      <td>MYO6</td>\n",
       "      <td>600970</td>\n",
       "      <td>7605</td>\n",
       "      <td>DFNA22; DFNB37; KIAA0389</td>\n",
       "      <td>MYO6-related nonsyndromic genetic hearing loss...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0019497</td>\n",
       "      <td>monoallelic_autosomal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>moderate</td>\n",
       "      <td>decreased gene product level; altered gene pro...</td>\n",
       "      <td>splice_donor_variant; splice_acceptor_variant;...</td>\n",
       "      <td>undetermined</td>\n",
       "      <td>inferred</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30681346; 24105371; 15060111; 12687499; 183482...</td>\n",
       "      <td>Cardiac</td>\n",
       "      <td>There is evidence from multiple reports that l...</td>\n",
       "      <td>2024-10-04 11:33:50+00:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>G2P03449</td>\n",
       "      <td>MYO6</td>\n",
       "      <td>600970</td>\n",
       "      <td>7605</td>\n",
       "      <td>DFNA22; DFNB37; KIAA0389</td>\n",
       "      <td>MYO6-related nonsyndromic genetic hearing loss...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0019497</td>\n",
       "      <td>biallelic_autosomal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>moderate</td>\n",
       "      <td>decreased gene product level; altered gene pro...</td>\n",
       "      <td>splice_donor_variant; splice_acceptor_variant;...</td>\n",
       "      <td>undetermined</td>\n",
       "      <td>inferred</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30681346; 24105371; 15060111; 12687499; 183482...</td>\n",
       "      <td>Cardiac</td>\n",
       "      <td>There is evidence from multiple reports that l...</td>\n",
       "      <td>2024-10-04 11:33:37+00:00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      g2p id gene symbol  gene mim  hgnc id  \\\n",
       "2   G2P03247        DSC2    125645     3036   \n",
       "3   G2P03248        DSC2    125645     3036   \n",
       "4   G2P03249        DSG2    125671     3049   \n",
       "5   G2P03250        DSG2    125671     3049   \n",
       "6   G2P03251         DSP    125647     3052   \n",
       "7   G2P03252         DSP    125647     3052   \n",
       "17  G2P03262       CASQ2    114251     1513   \n",
       "18  G2P03263       CASQ2    114251     1513   \n",
       "40  G2P03285       KCNQ1    607542     6294   \n",
       "41  G2P03286       KCNQ1    607542     6294   \n",
       "44  G2P03289       ALPK3    617608    17574   \n",
       "45  G2P03290       ALPK3    617608    17574   \n",
       "46  G2P03291         DES    125660     2770   \n",
       "47  G2P03292         DES    125660     2770   \n",
       "70  G2P03446        MYO6    600970     7605   \n",
       "73  G2P03449        MYO6    600970     7605   \n",
       "\n",
       "                            previous gene symbols  \\\n",
       "2                                     CDHF2; DSC3   \n",
       "3                                     CDHF2; DSC3   \n",
       "4                                           CDHF5   \n",
       "5                                           CDHF5   \n",
       "6                         DPI; DPII; KPPS2; PPKS2   \n",
       "7                         DPI; DPII; KPPS2; PPKS2   \n",
       "17                                          PDIB2   \n",
       "18                                          PDIB2   \n",
       "40  JLNS1; KCNA8; KCNA9; KV7.1; KVLQT1; LQT; LQT1   \n",
       "41  JLNS1; KCNA8; KCNA9; KV7.1; KVLQT1; LQT; LQT1   \n",
       "44                          KIAA1330; MAK; MIDORI   \n",
       "45                          KIAA1330; MAK; MIDORI   \n",
       "46                      CMD1I; CSM1; CSM2; LGMD2R   \n",
       "47                      CMD1I; CSM1; CSM2; LGMD2R   \n",
       "70                       DFNA22; DFNB37; KIAA0389   \n",
       "73                       DFNA22; DFNB37; KIAA0389   \n",
       "\n",
       "                                         disease name  disease mim  \\\n",
       "2   DSC2-related arrhythmogenic right ventricular ...          NaN   \n",
       "3   DSC2-related arrhythmogenic right ventricular ...          NaN   \n",
       "4   DSG2-related arrhythmogenic right ventricular ...          NaN   \n",
       "5   DSG2-related arrhythmogenic right ventricular ...          NaN   \n",
       "6   DSP-related arrhythmogenic right ventricular c...          NaN   \n",
       "7   DSP-related arrhythmogenic right ventricular c...          NaN   \n",
       "17  CASQ2-related catecholaminergic polymorphic ve...          NaN   \n",
       "18  CASQ2-related catecholaminergic polymorphic ve...          NaN   \n",
       "40                     KCNQ1-related long QT syndrome          NaN   \n",
       "41                     KCNQ1-related long QT syndrome          NaN   \n",
       "44          ALPK3-related hypertrophic cardiomyopathy          NaN   \n",
       "45          ALPK3-related hypertrophic cardiomyopathy          NaN   \n",
       "46                  DES-related myofibrillar myopathy          NaN   \n",
       "47                  DES-related myofibrillar myopathy          NaN   \n",
       "70  MYO6-related nonsyndromic genetic hearing loss...          NaN   \n",
       "73  MYO6-related nonsyndromic genetic hearing loss...          NaN   \n",
       "\n",
       "    disease MONDO    allelic requirement cross cutting modifier  confidence  \\\n",
       "2   MONDO:0012506  monoallelic_autosomal                    NaN  definitive   \n",
       "3   MONDO:0012506    biallelic_autosomal                    NaN  definitive   \n",
       "4   MONDO:0012434  monoallelic_autosomal                    NaN  definitive   \n",
       "5   MONDO:0012434    biallelic_autosomal                    NaN  definitive   \n",
       "6   MONDO:0011831  monoallelic_autosomal                    NaN  definitive   \n",
       "7   MONDO:0011831    biallelic_autosomal                    NaN  definitive   \n",
       "17  MONDO:0012762  monoallelic_autosomal                    NaN    moderate   \n",
       "18  MONDO:0012762    biallelic_autosomal                    NaN  definitive   \n",
       "40  MONDO:0100316  monoallelic_autosomal                    NaN  definitive   \n",
       "41  MONDO:0100316    biallelic_autosomal                    NaN  definitive   \n",
       "44  MONDO:0054838  monoallelic_autosomal                    NaN      strong   \n",
       "45  MONDO:0054838    biallelic_autosomal                    NaN      strong   \n",
       "46  MONDO:0011076  monoallelic_autosomal                    NaN  definitive   \n",
       "47  MONDO:0011076    biallelic_autosomal                    NaN  definitive   \n",
       "70  MONDO:0019497  monoallelic_autosomal                    NaN    moderate   \n",
       "73  MONDO:0019497    biallelic_autosomal                    NaN    moderate   \n",
       "\n",
       "                                  variant consequence  \\\n",
       "2   decreased gene product level; altered gene pro...   \n",
       "3   decreased gene product level; altered gene pro...   \n",
       "4   decreased gene product level; altered gene pro...   \n",
       "5   decreased gene product level; altered gene pro...   \n",
       "6   decreased gene product level; altered gene pro...   \n",
       "7   decreased gene product level; altered gene pro...   \n",
       "17  decreased gene product level; altered gene pro...   \n",
       "18  absent gene product; altered gene product stru...   \n",
       "40  decreased gene product level; altered gene pro...   \n",
       "41  decreased gene product level; altered gene pro...   \n",
       "44  decreased gene product level; altered gene pro...   \n",
       "45  decreased gene product level; altered gene pro...   \n",
       "46  absent gene product; altered gene product stru...   \n",
       "47  absent gene product; altered gene product stru...   \n",
       "70  decreased gene product level; altered gene pro...   \n",
       "73  decreased gene product level; altered gene pro...   \n",
       "\n",
       "                                        variant types molecular mechanism  \\\n",
       "2   splice_region_variant; inframe_insertion; spli...        undetermined   \n",
       "3   splice_region_variant; inframe_insertion; spli...        undetermined   \n",
       "4   inframe_insertion; splice_donor_variant; infra...        undetermined   \n",
       "5   inframe_insertion; splice_donor_variant; infra...        undetermined   \n",
       "6   inframe_insertion; splice_donor_variant; infra...        undetermined   \n",
       "7   inframe_insertion; splice_donor_variant; infra...        undetermined   \n",
       "17  missense_variant; splice_region_variant; splic...        undetermined   \n",
       "18  missense_variant; splice_region_variant; splic...        undetermined   \n",
       "40  frameshift_variant_NMD_triggering; inframe_ins...        undetermined   \n",
       "41  frameshift_variant_NMD_triggering; inframe_ins...        undetermined   \n",
       "44  frameshift_variant_NMD_triggering; splice_dono...        undetermined   \n",
       "45  frameshift_variant_NMD_triggering; splice_dono...        undetermined   \n",
       "46  frameshift_variant_NMD_triggering; splice_dono...        undetermined   \n",
       "47  frameshift_variant_NMD_triggering; splice_dono...        undetermined   \n",
       "70  splice_donor_variant; splice_acceptor_variant;...        undetermined   \n",
       "73  splice_donor_variant; splice_acceptor_variant;...        undetermined   \n",
       "\n",
       "   molecular mechanism categorisation  molecular mechanism evidence  \\\n",
       "2                            inferred                           NaN   \n",
       "3                            inferred                           NaN   \n",
       "4                            inferred                           NaN   \n",
       "5                            inferred                           NaN   \n",
       "6                            inferred                           NaN   \n",
       "7                            inferred                           NaN   \n",
       "17                           inferred                           NaN   \n",
       "18                           inferred                           NaN   \n",
       "40                           inferred                           NaN   \n",
       "41                           inferred                           NaN   \n",
       "44                           inferred                           NaN   \n",
       "45                           inferred                           NaN   \n",
       "46                           inferred                           NaN   \n",
       "47                           inferred                           NaN   \n",
       "70                           inferred                           NaN   \n",
       "73                           inferred                           NaN   \n",
       "\n",
       "   phenotypes                                       publications    panel  \\\n",
       "2         NaN  31028357; 23911551; 21636032; 33831308; 263105...  Cardiac   \n",
       "3         NaN  31028357; 23911551; 21636032; 33831308; 263105...  Cardiac   \n",
       "4         NaN  21636032; 33831308; 33917638; 34400560; 240707...  Cardiac   \n",
       "5         NaN  21636032; 33831308; 33917638; 34400560; 240707...  Cardiac   \n",
       "6         NaN  32372669; 23137101; 21636032; 33831308; 313199...  Cardiac   \n",
       "7         NaN  32372669; 23137101; 21636032; 33831308; 313199...  Cardiac   \n",
       "17        NaN  12386154; 16908766; 21618644; 15176429; 326936...  Cardiac   \n",
       "18        NaN  12386154; 16908766; 21618644; 15176429; 326936...  Cardiac   \n",
       "40        NaN  31983240; 9020846; 23392653; 23591039; 2704115...  Cardiac   \n",
       "41        NaN  31983240; 9020846; 23392653; 23591039; 2704115...  Cardiac   \n",
       "44        NaN  27106955; 31074094; 32480058; 26846950; 214411...  Cardiac   \n",
       "45        NaN  27106955; 31074094; 32480058; 26846950; 214411...  Cardiac   \n",
       "46        NaN  30681346; 19181099; 20718792; 29926427; 194333...  Cardiac   \n",
       "47        NaN  30681346; 19181099; 20718792; 29926427; 194333...  Cardiac   \n",
       "70        NaN  30681346; 24105371; 15060111; 12687499; 183482...  Cardiac   \n",
       "73        NaN  30681346; 24105371; 15060111; 12687499; 183482...  Cardiac   \n",
       "\n",
       "                                             comments  \\\n",
       "2   Expert review done on 05/01/2022; DSC2-related...   \n",
       "3   Expert review done on 05/01/2022; DSC2-related...   \n",
       "4   Expert review done on 05/01/2022; DSG2-related...   \n",
       "5   Expert review done on 05/01/2022; DSG2-related...   \n",
       "6   Expert review done on 05/01/2022; DSP-related ...   \n",
       "7   Expert review done on 05/01/2022; DSP-related ...   \n",
       "17  Expert review done on 19/05/2021; CASQ2 is an ...   \n",
       "18  Expert review done on 19/05/2021; CASQ2 is an ...   \n",
       "40  Expert review done on 16/06/2021; Pathogenic v...   \n",
       "41  Expert review done on 16/06/2021; Pathogenic v...   \n",
       "44  Expert review done on 24/11/2021; ALPK3 pathog...   \n",
       "45  Expert review done on 24/11/2021; ALPK3 pathog...   \n",
       "46  Expert review done on 12/01/2022; DES-related ...   \n",
       "47  Expert review done on 12/01/2022; DES-related ...   \n",
       "70  There is evidence from multiple reports that l...   \n",
       "73  There is evidence from multiple reports that l...   \n",
       "\n",
       "          date of last review  \n",
       "2   2024-03-20 09:36:09+00:00  \n",
       "3   2024-03-20 09:35:19+00:00  \n",
       "4   2024-03-20 09:40:18+00:00  \n",
       "5   2024-03-20 09:37:58+00:00  \n",
       "6   2024-03-20 09:42:53+00:00  \n",
       "7   2024-03-20 09:42:01+00:00  \n",
       "17  2024-03-13 11:49:22+00:00  \n",
       "18  2022-04-25 12:45:21+00:00  \n",
       "40  2024-03-13 11:56:22+00:00  \n",
       "41  2024-03-13 11:57:14+00:00  \n",
       "44  2022-04-25 12:45:22+00:00  \n",
       "45  2022-04-25 12:45:22+00:00  \n",
       "46  2022-04-25 12:45:22+00:00  \n",
       "47  2024-03-26 10:31:28+00:00  \n",
       "70  2024-10-04 11:33:50+00:00  \n",
       "73  2024-10-04 11:33:37+00:00  "
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## there's 1 gene-disease pair duplicate\n",
    "cardiac[cardiac.duplicated(subset=['gene symbol', 'disease name'], keep=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "494cd482",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-03-17T19:56:22.837897Z",
     "start_time": "2025-03-17T19:56:22.829196Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 80 entries, 0 to 79\n",
      "Data columns (total 21 columns):\n",
      " #   Column                              Non-Null Count  Dtype  \n",
      "---  ------                              --------------  -----  \n",
      " 0   g2p id                              80 non-null     object \n",
      " 1   gene symbol                         80 non-null     object \n",
      " 2   gene mim                            80 non-null     object \n",
      " 3   hgnc id                             80 non-null     object \n",
      " 4   previous gene symbols               74 non-null     object \n",
      " 5   disease name                        80 non-null     object \n",
      " 6   disease mim                         6 non-null      object \n",
      " 7   disease MONDO                       73 non-null     object \n",
      " 8   allelic requirement                 80 non-null     object \n",
      " 9   cross cutting modifier              22 non-null     object \n",
      " 10  confidence                          80 non-null     object \n",
      " 11  variant consequence                 80 non-null     object \n",
      " 12  variant types                       79 non-null     object \n",
      " 13  molecular mechanism                 80 non-null     object \n",
      " 14  molecular mechanism categorisation  80 non-null     object \n",
      " 15  molecular mechanism evidence        0 non-null      float64\n",
      " 16  phenotypes                          14 non-null     object \n",
      " 17  publications                        80 non-null     object \n",
      " 18  panel                               80 non-null     object \n",
      " 19  comments                            80 non-null     object \n",
      " 20  date of last review                 80 non-null     object \n",
      "dtypes: float64(1), object(20)\n",
      "memory usage: 13.3+ KB\n"
     ]
    }
   ],
   "source": [
    "cardiac.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "0b278923",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-03-17T19:37:39.988735Z",
     "start_time": "2025-03-17T19:37:39.983718Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "80"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cardiac[\"g2p id\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "ed92489c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-03-17T19:59:24.880908Z",
     "start_time": "2025-03-17T19:59:24.874263Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "61"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "array([nan, 'MONDO:0012506', 'MONDO:0012434', 'MONDO:0011831',\n",
       "       'MONDO:0012180', 'MONDO:0011459', 'MONDO:0011017', 'MONDO:0013966',\n",
       "       'MONDO:0017990', 'MONDO:0013529', 'MONDO:0014191', 'MONDO:0011001',\n",
       "       'MONDO:0012762', 'MONDO:0012312', 'MONDO:0012314', 'MONDO:0012313',\n",
       "       'MONDO:0000453', 'MONDO:0013479', 'MONDO:0011482', 'MONDO:0005021',\n",
       "       'MONDO:0007269', 'MONDO:0013262', 'MONDO:0012362', 'MONDO:0013168',\n",
       "       'MONDO:0012745', 'MONDO:0007267', 'MONDO:0011400', 'MONDO:0012799',\n",
       "       'MONDO:0007268', 'MONDO:0012112', 'MONDO:0012111', 'MONDO:0013369',\n",
       "       'MONDO:0013367', 'MONDO:0100316', 'MONDO:0011377', 'MONDO:0019171',\n",
       "       'MONDO:0054838', 'MONDO:0011076', 'MONDO:0010680', 'MONDO:0012289',\n",
       "       'MONDO:0010526', 'MONDO:0010281', 'MONDO:0010946', 'MONDO:0024540',\n",
       "       'MONDO:0008919', 'MONDO:0008647', 'MONDO:0014548', 'MONDO:0014550',\n",
       "       'MONDO:0007266', 'MONDO:0008222', 'MONDO:0010979', 'MONDO:0008104',\n",
       "       'MONDO:0012690', 'MONDO:0014143', 'MONDO:0100339', 'MONDO:0019497',\n",
       "       'MONDO:0044970', 'MONDO:0018943', 'MONDO:0011581', 'MONDO:0011003',\n",
       "       'MONDO:0019441', 'MONDO:0013195'], dtype=object)"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cardiac[\"disease MONDO\"].nunique()\n",
    "cardiac[\"disease MONDO\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "77457e21",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-03-17T20:00:03.147351Z",
     "start_time": "2025-03-17T20:00:03.141554Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4, 21)"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# cardiac[cardiac[\"disease mim\"] == 612347]\n",
    "\n",
    "cardiac[cardiac[\"disease mim\"].isna() & cardiac[\"disease MONDO\"].isna()].shape\n",
    "\n",
    "# cancer[cancer[\"comments\"].notna()]\n",
    "\n",
    "# cardiac[\"disease mim\"].astype(str).unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0323f99",
   "metadata": {},
   "source": [
    "### Organizing data into documents/individual association objects:\n",
    "\n",
    "#### NOW ON: \n",
    "cardiac: allelic requirement\n",
    "\n",
    "\n",
    "#### Gene subject section:\n",
    "- \"gene symbol\": str. \"HGNC-assigned gene symbol\" (according to DataDownloadFormat txt file)\n",
    "  - DOC TRANSFORM: hgnc_symbol\n",
    "- \"gene mim\": OMIM ID for gene. NodeNorm does recognize these\n",
    "  - DF READ: force dtype to be str\n",
    "  - DF CHECK: \n",
    "    - Unsure: add current Translator prefix\n",
    "  - DOC TRANSFORM: omim\n",
    "- \"hgnc id\"\n",
    "  - DF READ: force dtype to be str\n",
    "  - DF CHECK: same as \"gene mim\"\n",
    "- \"previous gene symbols\": \"; \"-delimited.\n",
    "  - DOC TRANSFORM:\n",
    "    - CHECK for NA. Only assign field if not NA\n",
    "    - Turn into array of str: \";\"-split and then strip whitespace \n",
    "\n",
    "#### Disease object section:\n",
    "- \"disease name\":\n",
    "  - DOC TRANSFORM: name\n",
    "- \"disease mim\": OMIM ID for disease. But there are some Orphanet IDs mixed in (start with \"Orphanet:\")\n",
    "  - DF READ: force dtype to be str\n",
    "  - DF CHECK:\n",
    "    - Unsure: add/replace current Translator prefix\n",
    "  - DOC TRANSFORM: omim, orphanet\n",
    "    - CHECK for NA (many rows are). Only assign field if not NA\n",
    "    - CHECK for \"Orphanet:\": if yes, assign to orphanet field instead\n",
    "- \"disease MONDO\": MONDO ID for disease (already Translator-formatted CURIE) \n",
    "  - DOC TRANSFORM: mondo\n",
    "    - CHECK for NA (many rows are). Only assign field if not NA\n",
    "    \n",
    "#### Association section:\n",
    "- \"g2p id\": resource's unique stable ID for each record/row (according to DataDownloadFormat txt file, G2P webpage for single record)\n",
    "  - DF TRANSFORM:\n",
    "    - create new column, generating G2P webpage urls for each record like https://www.ebi.ac.uk/gene2phenotype/lgd/G2P03386. this may be brittle/work to maintain, but I know UI really wants stuff like this\n",
    "  - DOC TRANSFORM: g2p_record_id, g2p_record_url\n",
    "    - use to create\n",
    "- \"allelic requirement\": required genotype (according to G2P webpage for single record). str (categorical). [G2P website](https://www.ebi.ac.uk/gene2phenotype/about/terminology) says these are synonyms of HPO \"mode of inheritance terms\", which seem to be from the [inheritance part of the ontology](https://hpo.jax.org/browse/term/HP:0000005). Ive included a mapping table in reference section.\n",
    "  - DF TRANSFORM:\n",
    "    - optional: turn dtype into 'category'\n",
    "    - very optional: create new columns for HPO mapping. Probably based on static mapping dict, which would be brittle/work to maintain\n",
    "- \"cross cutting modifier\": additional info relevant to gene-disease inheritance (according to G2P webpage for single record). Can be \"; \"-delimited, otherwise would be categorical. [G2P website](https://www.ebi.ac.uk/gene2phenotype/about/terminology) says these are HPO inheritance qualifier terms \"when available\" ([this part of ontology](https://hpo.jax.org/browse/term/HP:0034335)). I didn't try mapping these.\n",
    "  - DOC TRANSFORM:\n",
    "    - CHECK for NA (many rows are). Only assign field if not NA\n",
    "    - Turn into array of str: \";\"-split and then strip whitespace \n",
    "- \"confidence\": confidence that the association is real (according to G2P webpage for panel). [G2P website](https://www.ebi.ac.uk/gene2phenotype/about/terminology) has definitions of terms. str (categorical)\n",
    "  - DF TRANSFORM:\n",
    "    - optional: turn dtype into 'category'\n",
    "- \"variant consequence\": consequences of reported variants on product (protein or RNA) (according to G2P webpage for single record). Can be \"; \"-delimited, otherwise would be categorical. [G2P website](https://www.ebi.ac.uk/gene2phenotype/about/terminology) has definitions of terms and SO mappings. \n",
    "  - DF TRANSFORM:\n",
    "    - very optional: create new columns for SO mapping. Probably based on static mapping dict, which would be brittle/work to maintain\n",
    "  - DOC TRANSFORM:\n",
    "    - Turn into array of str: \";\"-split and then strip whitespace \n",
    "- \"variant types\": associated with gene-disease pair (according to G2P webpage for single record). Can be \"; \"-delimited, otherwise would be categorical. [G2P website](https://www.ebi.ac.uk/gene2phenotype/about/terminology) has definitions of terms and SO mappings (large table!). \n",
    "  - DF TRANSFORM:\n",
    "    - very optional: create new columns for SO mapping. Probably based on static mapping dict, which would be brittle/work to maintain\n",
    "  - DOC TRANSFORM:\n",
    "    - CHECK for NA (many rows are). Only assign field if not NA\n",
    "    - Turn into array of str: \";\"-split and then strip whitespace \n",
    "- \"molecular mechanism\": mechanism of how gene's variants causes disease (according to G2P webpage for single record). str (categorical). [G2P website](https://www.ebi.ac.uk/gene2phenotype/about/terminology) has definitions of terms.\n",
    "  - DF TRANSFORM:\n",
    "    - optional: turn dtype into 'category'\n",
    "- \"molecular mechanism categorisation\": seems to say how molecular mechanism was decided. Shows up as \"support\" on G2P webpage for single record. possible values: \"inferred\" or \"evidence\". str (categorical). \n",
    "  - DF TRANSFORM:\n",
    "    - optional: turn dtype into 'category'\n",
    "- \"molecular mechanism evidence\": complicated structure. Looks like \" & \"-delimited, then primary_key (publication ID) -> object (diff key: value pairs split by \"; \", values can be \", \"-delimited lists). \n",
    "  - DOC TRANSFORM:\n",
    "    - CHECK for NA (many rows are). Only assign field if not NA\n",
    "    - Turn into array of str: \"&\"-split and then strip whitespace. Not going to do more parsing for now.  \n",
    "- \"phenotypes\": HPO IDs (already Translator-formatted CURIE) for phenotypes reported by the publications (according to G2P webpage for single record). Can be \"; \"-delimited\n",
    "  - based on G2P website, these are organized/assigned to specific gene-disease records...so I'm keeping as a part of the gene-disease assocation, rather than creating diff objects for gene-phenotype (if I did this, it would be a separate function)\n",
    "  - DOC TRANSFORM:\n",
    "    - CHECK for NA. Only assign field if not NA\n",
    "    - Turn into array of str: \";\"-split and then strip whitespace \n",
    "- \"publications\": these are PMIDs with no prefix (based on G2P webpage for single record). Can be \"; \"-delimited\n",
    "  - DOC TRANSFORM: pmids\n",
    "    - CHECK for NA. Only assign field if not NA\n",
    "    - Turn into array of str: \";\"-split and then strip whitespace \n",
    "- \"panel\": the G2P panels this record is assigned to (according to DataDownloadFormat txt file). Can be \"; \"-delimited, otherwise would be categorical.\n",
    "  - DOC TRANSFORM: g2p_panels\n",
    "    - map (str.replace) values to match G2P webpage for single record\n",
    "      - \"DD\" = \"Developmental disorders\"\n",
    "      - add \" disorders\" to end: \"Cancer\", \"Eye\", \"Skeletal\", \"Skin\",\n",
    "    - Turn into array of str: \";\"-split and then strip whitespace \n",
    "- \"comments\": additional comments from curation team (according to DataDownloadFormat txt file). Rare, appears to be free-text.\n",
    "  - DOC TRANSFORM: curator_comments\n",
    "    - CHECK for NA (many rows are). Only assign field if not NA\n",
    "- \"date of last review\":\n",
    "  - DF TRANSFORM:\n",
    "    - optional: turn dtype into datetime https://pandas.pydata.org/docs/reference/api/pandas.to_datetime.html#pandas.to_datetime\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#### Don't use:\n",
    "\n",
    "\n",
    "\n",
    "### Reference\n",
    "\n",
    "#### Allelic requirement terms\n",
    "| G2P | HPO name | HPO ID | Exact synonym? | \n",
    "| :- | :- | :- | :- |\n",
    "| biallelic_autosomal | Autosomal recessive inheritance | HP:0000007 | Yes |\n",
    "| monoallelic_autosomal | Autosomal dominant inheritance | HP:0000006 | Yes |\n",
    "| biallelic_PAR | Pseudoautosomal recessive inheritance | HP:0034341 | Yes |\n",
    "| monoallelic_PAR | Pseudoautosomal dominant inheritance | HP:0034340 | Yes |\n",
    "| mitochondrial | Mitochondrial inheritance | HP:0001427 | Yes |\n",
    "| monoallelic_Y_hemizygous | Y-linked inheritance | HP:0001450 | Yes |\n",
    "| monoallelic_X | X-linked inheritance | HP:0001417 | No (my guess) |\n",
    "| monoallelic_X_hemizygous | X-linked recessive inheritance | HP:0001419 | No (my guess) |\n",
    "| monoallelic_X_heterozygous | X-linked dominant inheritance | HP:0001423 | No (my guess) |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4638823f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d1d3577d",
   "metadata": {},
   "source": [
    "General Ideas:\n",
    "* each row = unique gene-disease-allelic requirement combo => turn into 1 association object\n",
    "  * \"G2P records are Locus-Genotype-Mechanism-Disease-Evidence (LGMDE) threads describing gene-disease associations\" -> this is what a unique row is from resource POV (ref: homepage table \"Total LGMDE Records\" info (i) button)\n",
    "* sometimes disease doesn't have any IDs for it => still make document/object but won't use in Translator unless it has ID\n",
    "* confidence is important to filter on for Translator! => will still make document/object but don't use! \n",
    "  * don't use \"limited\", \"disputed\", \"refuted\": Translator doesn't handle negation (\"refuted\") or low/conflicting evidence (\"disputed\" or \"limited\")\n",
    "\n",
    "\n",
    "\n",
    "  \n",
    "* if I loaded all data into 1 dataframe, there will be duplicate rows because the gene-disease pair is in several panels (aka disease falls into multiple categories). See how panel is a comma-delimited string => drop duplicates will be needed (but be careful to include all fields that matter for unique row)\n",
    "* str split method will create single-element arrays if no delimiter found"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a234ca0a",
   "metadata": {},
   "source": [
    "## BioThings Parser requirements"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5e31c9b",
   "metadata": {},
   "source": [
    "* document/object MUST have `_id` key (primary ID for BioThings database - ElasticSearch / mongo DB?) \n",
    "\n",
    "* parser feeds into upload step -> function should be generator to iterate through all stuff\" -> for each, yield the document/object as json\n",
    "\n",
    "\n",
    "\n",
    "* output needs to be json dump"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8734a06a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "220.5px"
   },
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
